### PR TITLE
Combat level is now a reflection of trained skills instead of accumul…

### PR DESCRIPTION
**Pull request by madsboddum**
…ated XP #385

### CODE REVIEW

**Comment by Undercova** [comment]
Working as intended

File: src/main/java/com/projectswg/holocore/services/gameplay/player/experience/skills/SkillService.java [hunk 1]
```diff
@@ -92,10 +146,25 @@ private void handleSurrenderSkillIntent(SurrenderSkillIntent ssi) {
...
+		
+		CombatLevel oldCombatLevel = getCombatLevel(target);
+		
 		target.removeSkill(surrenderedSkill);
 		target.removeCommands(skillInfo.getCommands());
 		skillInfo.getSkillMods().forEach((skillModName, skillModValue) -> new SkillModIntent(skillModName, 0, -skillModValue, target).broadcast());
+		
+		CombatLevel newCombatLevel = getCombatLevel(target);
+		
+		boolean levelChanged = !oldCombatLevel.equals(newCombatLevel);
```
Comment (Josh-Larson) [inline src/main/java/com/projectswg/holocore/services/gameplay/player/experience/skills/SkillService.java:158]: Is this variable necessary?

**Review by Josh-Larson — CHANGES_REQUESTED** [review]

File: src/main/java/com/projectswg/holocore/resources/support/data/server_info/loader/PlayerLevelLoader.java [hunk 2]
```diff
@@ -29,83 +29,56 @@
...
-	public int getMaxLevel() {
-		return maxLevel;
+		this.levels = new LinkedList<>();
 	}
 	
-	@Nullable
-	public PlayerLevelInfo getFromLevel(int level) {
-		return levels.get(level);
+	public Collection<PlayerLevelInfo> getPlayerLevelInfos() {
+		return levels;
```
[the remark this answers is not in the document]
Comment (madsboddum) [reply src/main/java/com/projectswg/holocore/resources/support/data/server_info/loader/PlayerLevelLoader.java:47]: Right, it's iterated now over to find the level that has the most fitting xp requirement.

In that way a map made less sense than a collection imo.

File: src/main/java/com/projectswg/holocore/services/gameplay/player/experience/ExperiencePointService.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		CreatureObject creatureObject = ei.getCreatureObject();
+		PlayerObject playerObject = creatureObject.getPlayerObject();
+		
+		if (playerObject != null) {
+			int experienceGained = ei.getExperienceGained();
+			String xpType = ei.getXpType();
+			
+			CombatXpMultiplierLoader combatXpMultiplierLoader = DataLoader.Companion.combatXpMultipliers();
+			CombatXpMultiplierLoader.CombatXpMultiplierInfo combatXpMultiplierInfo = combatXpMultiplierLoader.getCombatXpMultiplier(xpType);
+			boolean xpTypeAlsoGrantsCombatXp = combatXpMultiplierInfo != null;
```
[the remark this answers is not in the document]
Comment (madsboddum) [reply src/main/java/com/projectswg/holocore/services/gameplay/player/experience/ExperiencePointService.java:41]: I felt a named variable was the best way to convey what the following if-statementbdoes

File: src/main/java/com/projectswg/holocore/services/gameplay/player/experience/skills/SkillService.java (hunk 1 above)
Comment (madsboddum) [reply src/main/java/com/projectswg/holocore/services/gameplay/player/experience/skills/SkillService.java:158]: Same as above, it reads better imo

**Pull request merged**